src/rust/pants_ng/options/src/config.rs

```diff
@@ -0,0 +1,150 @@
+// Copyright 2025 Pants project contributors (see CONTRIBUTORS.md).
+// Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+use options::BuildRoot;
+use options::ConfigSource;
+use options::config::Config;
+use parking_lot::Mutex;
+use std::collections::HashMap;
+use std::collections::HashSet;
+use std::fs::canonicalize;
+use std::path::MAIN_SEPARATOR_STR;
+use std::path::Path;
+use std::path::PathBuf;
+use std::sync::Arc;
+
+pub type InterpolationMap = HashMap<String, String>;
+
+pub struct ConfigFinder {
+    buildroot: BuildRoot,
+    seed_values: InterpolationMap,
+    // Populated lazily as we encounter config files.
+    parsed_configs: Arc<Mutex<HashMap<PathBuf, Arc<Config>>>>,
+}
+
+impl ConfigFinder {
+    pub fn new(
+        buildroot: Option<BuildRoot>,
+        seed_values: InterpolationMap,
+    ) -> Result<Self, String> {
+        let buildroot = buildroot.unwrap_or(BuildRoot::find()?);
+
+        Ok(Self {
+            buildroot,
+            seed_values,
+            parsed_configs: Arc::new(Mutex::new(HashMap::new())),
+        })
+    }
+
+    fn get_config(&self, path: &Path) -> Result<Arc<Config>, String> {
+        let mut lock = self.parsed_configs.lock();
+        if let Some(config) = lock.get(path) {
+            return Ok(Arc::clone(config));
+        }
+        let config = Config::parse(&ConfigSource::from_file(path)?, &self.seed_values)?;
+        let ret = Arc::new(config);
+        lock.insert(path.to_path_buf(), Arc::clone(&ret));
+        Ok(ret)
+    }
+
+    // Returns all configs applicable to the given dir and context.
+    //
+    // See get_applicable_config_files() for how applicability is determined, and for
+    // the precedence order of the returned configs.
+    pub fn get_applicable_configs(
+        &self,
+        dir: &Path,
+        context: &HashSet<String>,
+    ) -> Result<Vec<Arc<Config>>, String> {
+        self.get_applicable_config_files(dir, context)?
+            .into_iter()
+            .map(|path| self.get_config(path.as_path()))
+            .collect()
+    }
+
+    // Returns all config files applicable to the given dir and context.
+    //
+    // That is, all config files in that dir and in all ancestor dirs up to the build root,
+    // all of whose context tags are in the specified context.
+    //
+    // A valid config file name is `pants.toml` or `pants_(context tags).toml` where the context
+    // tags are dot-separated strings.
+    //
+    // For example, "pants.ci.macos.arm64.toml" has the context tags "ci", "macos" and "arm64",
```

Comment:
I kind of like the period better than underscore. Did that cause any ambiguity?

```suggestion
    // For example, "pants_ci.macos.arm64.toml" has the context tags "ci", "macos" and "arm64",
```

Another point: at least Konsole and probably other tools that `_` as a word character and `.` or `-` are word breaking characters. So, double clicking ci would select `ci` with `.` or `-` but would select `pants_ci` with `_`.